38863 Update Engineering Diffraction Calibration region of interest functionality

andy-bridger opened the pull request:
### Description of work

#### Summary of work
Added support so the Custom File provided to Engineering Diffraction Calibration can be an `xml`, such as those output by `SaveDetectorsGrouping`

Fixes #38863 . <!-- and fix #xxxx or close #xxxx xor resolves #xxxx. One line per issue fixed. -->
Also fixes a bug I found in testing new implementation #38879

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

#### Further detail of work
Changed it so the interface now accepts `xml` files of the grouping. Additionally changed both the user facing and internal reference to this from `CalFile` (or variations upon this) to `GroupingFile` and in the UI `Crop Calibration` to `Set Calibration Region of Interest`, these provide more accurate names to actual functionality and should make the interface more intuitive to use.
Changed the naming suffix for custom file `example_group.xml` from `_Custom` to `_Custom_example_group` so they don't get overwritten when custom grouping is changed (also more clear to the user what grouping is being used).
Changed the naming suffix for custom crop `example_spec` from `_Cropped` to `_Cropped_example_spec` so they don't get overwritten when custom grouping is changed (also more clear to the user what grouping is being used).
Finally changed it for the vanadium normalisation so a warning is flagged if `GROUP.CUSTOM` or `GROUP.CROPPED` are loaded in from the ADS, which, together with more unique naming, should cover the bug #38879

### To test:

Open the Engineering Diffraction Interface

Select `Create New Calibration`

Set the sample as `305738`

Select `Set Calibration Region of Interest`

Set Region of Interest to `Custom Grouping File` in the drop down

Try setting a custom grouping in `xml` format (code to create one provided below)

Calibration should be successful 

Try using calibration to focus on the next tab 

If you use the customs below should get: 2 spectra for `"bank"`, 10 for `"module"`, and 90 for `"block"`

(Bank should give identical results to Test 1 in the manual testing: `https://developer.mantidproject.org/Testing/EngineeringDiffraction/EngineeringDiffractionTestGuide.html`)

```
from mantid.simpleapi import *

def get_detector_grouping_string(ws, group_by):
    info = ws.componentInfo()
    detinfo = ws.detectorInfo()
    dets = detinfo.detectorIDs()
    instr_dets = info.detectorsInSubtree(info.root())

    def get_det_id(comp_ind, dets, instr_dets):
        return dets[np.where(instr_dets == comp_ind)][0]

    nbi = info.indexOfAny("NorthBank")
    sbi = info.indexOfAny("SouthBank")


    nbmi = info.children(nbi)
    sbmi = info.children(sbi)

    nbmbi = [xx for x in [info.children(int(nbm)) for nbm in nbmi] for xx in x]
    sbmbi = [xx for x in [info.children(int(sbm)) for sbm in sbmi] for xx in x]
    if group_by == "bank":
        n_dets = "+".join([str(get_det_id(x, dets, instr_dets)) for x in info.detectorsInSubtree(nbi)])
        s_dets = "+".join([str(get_det_id(x, dets, instr_dets)) for x in info.detectorsInSubtree(sbi)])
        print([d for d in n_dets.split("+") if det_is_monitor(d, detinfo)])
        print([d for d in s_dets.split("+") if det_is_monitor(d, detinfo)])
        return ",".join([n_dets, s_dets])
    if group_by == "module":
        n_mods = ",".join(
            ["+".join([str(get_det_id(x, dets, instr_dets)) for x in info.detectorsInSubtree(int(nbm))]) for nbm in
             nbmi])
        s_mods = ",".join(
            ["+".join([str(get_det_id(x, dets, instr_dets)) for x in info.detectorsInSubtree(int(sbm))]) for sbm in
             sbmi])
        return ",".join([n_mods, s_mods])
    if group_by == "block":
        n_blocks = ",".join(
            ["+".join([str(get_det_id(x, dets, instr_dets)) for x in info.detectorsInSubtree(int(nbm))]) for nbm in
             nbmbi])
        s_blocks = ",".join(
            ["+".join([str(get_det_id(x, dets, instr_dets)) for x in info.detectorsInSubtree(int(sbm))]) for sbm in
             sbmbi])
        return ",".join([n_blocks, s_blocks])
        
ws = LoadNexus(r"C:\Users\kcd17618\Downloads\ENGINX00218189.nxs") #put any enginx run here

group_type = "module" #bank/module/block should all make different custom groupings

CreateGroupingWorkspace(ws,OutputWorkspace ="group_ws", 
                        CustomGroupingString=get_detector_grouping_string(ws, "module"), 
                        ComponentName = "ENGIN-X")
                        
SaveDetectorsGrouping("group_ws", f"custom_{group_type}.xml")
```

You can also run through the reproduction instructions here: #38879 and check that the bug is no longer there.

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

andy-bridger committed:
allow xml files to be loaded

andy-bridger committed:
refactored variable name's ref cal to grouping

andy-bridger committed:
added test coverage

added EnggUtils test coverage for custom and cropped groups

added spec string to Cropped suffix + tests

added tests for additional suffix logic

fixed test coverage for custom grouping behaviour

fixed test

andy-bridger committed:
refactor, type-hint and tidy code

remove unnecessary local variable

added type hinting and refactored repeated code

replace equivalent logic with ternary

andy-bridger committed:
update crop calibration to roi calibration

andy-bridger committed:
add unique custom group names

andy-bridger committed:
update release notes

updated note and added bugfix note

improved clarity of release notes

andy-bridger committed:
change van_focus rerun condition and add warning

bug fix focussing error when same calibration name

JackEAllen reviewed (commented):
I have not functionally tested yet, but from an initial pass through the proposed changes,  everything looks good! - only a few minor suggestions.
I wonder if it would be worthwhile updating `dev-docs/source/Testing/EngineeringDiffraction/EngineeringDiffractionTestGuide.rst` to add a test to validate that #38879 and #38863 are not accidentally reintroduced in the future. I feel like this would be a good smoke test.

andy-bridger commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/common/cropping/cropping_presenter.py:
```diff
@@ -19,58 +19,58 @@ def __init__(self, parent, view=None, model=None):
 
-    def on_calfile_changed(self):
+    def on_groupingfile_changed(self):
         valid = self.view.finder_custom.isValid()
```
Is there any reason to keep the local `valid` rather than just setting straight to `self.groupingfile_valid` and checking that in the subsequent logic?

andy-bridger replied:
`groupingfile_valid` is used in a couple of the tests, so probably end up being more code to remove it. I had just gotten rid of the local `valid`, but I haven't seen the python ternary operator outside of a list comp so will use that instead (and probably add it to a few other places that might apply)

andy-bridger commented on scripts/Engineering/common/calibration_info.py:
```diff
@@ -135,20 +160,25 @@ def set_calibration_from_prm_fname(self, file_path):
 
-    def set_cal_file(self, cal_filepath):
-        self.cal_filepath = cal_filepath
+    def set_grouping_file(self, grouping_filepath):
```
I have just added type hinting across the whole file, as it was a bit mix and match.

andy-bridger committed:
gui presenter updates cal better

sf1919 commented:
Genuine docs failure

andy-bridger commented:
A bit of a further update, especially re: manual testing.

I was reluctant to add a manual test as it felt like it could be automatically tested and I know people don't love doing the manual testing. I have added a system test that is essentially carrying out the procedure of running a calibration and focus with one custom group file, then running it again with a different group - and checking that this results in a different final spec. 

To get this working I did end up making a few more changes. Notably, adding the custom spectra string to the suffix. 

I also found that `update_calibration_from_view()` in the presenter wasn't updating the group workspaces which is a continuation of the initial bug report, but probably impacted more cases than I initially thought.

Re doc failure: I tried to change so the release notes link always has the algorithm name, but forgot to remove the `<>`

JackEAllen reviewed (changes requested):
Thanks for the changes and additional tests! A system test is a good idea to save on adding another smoke test. 

Would you be happy to rebase you're commits to squash them down? From looking through them, I think we can reduce them and possible re-order here and there. Here's a nice blog by GitHub I have used as a reference for structuring my own commits in the past: https://github.blog/developer-skills/github/write-better-commits-build-better-projects/
I've added a few minor suggestions to condense some methods slightly. but overall, your changes look really good and work as expected from manual testing.

andy-bridger committed:
tidy up of cropping presenter

andy-bridger committed:
improve clarity in calib info

andy-bridger commented:
This is my first time so aggressively messing with commit history, so I would appreciate it if you could give a final quick sanity check that I managed to do that successfully and haven't broken things...

JackEAllen commented:
> This is my first time so aggressively messing with commit history, so I would appreciate it if you could give a final quick sanity check that I managed to do that successfully and haven't broken things...

Not to worry! Happy to have a look. It can be scary the first couple of times

JackEAllen commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/common/cropping/cropping_presenter.py:
```diff
@@ -19,74 +19,67 @@ def __init__(self, parent, view=None, model=None):
 
     # Signal Activated Functions
 
     def on_combo_changed(self, index):
```
Thanks for adding this

andy-bridger committed:
remove try condition for extra suffix

RichardWaiteSTFC commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/common/cropping/cropping_presenter.py:
```diff
@@ -19,110 +19,84 @@ def __init__(self, parent, view=None, model=None):
+        self.group = group_mapping.get(index, GROUP.TEXTURE30)
+        self.custom_groupingfile_enabled = index == 0
+        self.custom_spectra_enabled = index == 3
+        self.set_custom_widgets_visibility(index == 0, index == 3)
```
Not sure if it's more readable but technically avoids some small duplication
```suggestion
        self.set_custom_widgets_visibility(self.custom_groupingfile_enabled, self.custom_spectra_enabled)
```

RichardWaiteSTFC commented on scripts/Engineering/common/calibration_info.py:
```diff
@@ -215,25 +252,43 @@ def get_subplot_title(self, ispec):
+            raise ValueError("Grouping file path is not set.")
+        ext = path.splitext(self.grouping_filepath)[-1].lower()
+        if ext == ".cal":
+            self.create_grouping_workspace_from_calfile()
```
Do we need a separate function for this - not sure why it was like that in the first place (probably me!)
```suggestion
            grp_ws, _, _ = CreateGroupingWorkspace(InstrumentName=self.instrument, OldCalFilename=self.grouping_filepath, OutputWorkspace=self.get_group_ws_name()
```

RichardWaiteSTFC commented on scripts/Engineering/EnggUtils.py:
```diff
@@ -273,13 +273,13 @@ def load_existing_calibration_files(calibration):
         msg = f"Could not open GSAS calibration file: {prm_filepath}"
         logger.warning(msg)
         return None
+    calibration.load_relevant_calibration_files()
```
This being moved here looks like it's important/a  bug fix - is that right? If so, is it worth a comment to say why it has to be done in this order?

andy-bridger replied:
I think I tried moving it when I was chasing the vanadium normalisation bug and just left it there. I can move it back if wanted

RichardWaiteSTFC replied:
No worries, happy to leave it where it is!

RichardWaiteSTFC commented on scripts/Engineering/EnggUtils.py:
```diff
@@ -506,6 +504,14 @@ def process_vanadium(vanadium_path, calibration, full_calib):
     van_foc_name = CURVES_PREFIX + calibration.get_group_suffix()
     if ADS.doesExist(van_foc_name):
+        if calibration.group == GROUP.CUSTOM or calibration.group == GROUP.CROPPED:
+            logger.warning(
```
Thansk for fixing this bug with the vanadium workspace naming. Is it worth checking the cached vanadium has the same number of histograms as the focussed workspace before dividing? In theory sone could also check the DIFCs within some tolerance if you were worried about different groupings with same number of spectra?

andy-bridger replied:
Having thought some more I think best is we keep the warning when the vanadium is reloaded and then before it is applied run the checks and if they fail then throw the error:

```
def _check_ws_foc_and_ws_van_foc(ws_foc, ws_van_foc):
    try:
        get_difc = lambda ws, ind : ws.spectrumInfo().diffractometerConstants(ind)[UnitParams.difc]
        num_foc, num_van = ws_foc.getNumberHistograms(), ws_van_foc.getNumberHistograms()
        assert num_foc == num_van
        assert get_difc(ws_foc, 0) == get_difc(ws_van_foc, 0)
        assert get_difc(ws_foc, num_foc-1) == get_difc(ws_van_foc, num_van-1)
    except AssertionError:
        logger.error(f"The calibration of {ws_van_foc} does not match {ws_foc}. "
                     f"Ensure the vanadium calibration file loaded is correct.")

```

I'll also add a system test to check this works

RichardWaiteSTFC replied:
Thanks, you could always add a unit test file just for EnggUtils?

RichardWaiteSTFC reviewed (changes requested):
Thanks for this, don't worry about the commit history it's now recommended we squash on merging so it will all go in a single commit anyway! Still it's good practice to do it as you go along as it helps the reviewer!

Just some super minor changes requested (@JackEAllen has done a very thorough review!). Thanks for updating the manual testing docs, we should also update the screenshots as well as you've changed the names/labels for some things on the calibration tab.

andy-bridger committed:
remove repeated code

andy-bridger committed:
add check before applying van norm + test

andy-bridger commented:
@JackEAllen I rebased to try and fix my unit tests so idk if I need you to reapprove?

@RichardWaiteSTFC I've added some more handling for if different groupings happen to have the same name and some System Tests to check it is working. Re: screenshots I don't think any need to be updated (https://developer.mantidproject.org/Testing/EngineeringDiffraction/EngineeringDiffractionTestGuide.html) as none show the tabs?

RichardWaiteSTFC commented:
> @JackEAllen I rebased to try and fix my unit tests so idk if I need you to reapprove?
> 
> @RichardWaiteSTFC I've added some more handling for if different groupings happen to have the same name and some System Tests to check it is working. Re: screenshots I don't think any need to be updated (https://developer.mantidproject.org/Testing/EngineeringDiffraction/EngineeringDiffractionTestGuide.html) as none show the tabs?

Thanks, there's some more user docs (not testing docs for devs) which has a screenshot (and might also need some text updating?)
https://github.com/mantidproject/mantid/blob/92b56c7484f016f35aa331077cd65913e1b33d98/docs/source/interfaces/diffraction/Engineering%20Diffraction.rst?plain=1#L123

andy-bridger committed:
update the eng diff interface docs

andy-bridger committed:
update testing files

andy-bridger committed:
fix focus_model unit test

RichardWaiteSTFC reviewed (commented):
Thanks for adding the DIFC check and the new systems tests look good. The sysem test failure is odd, it looks like it worked for one new file added
https://testdata.mantidproject.org/ftp/external-data/MD5/e6caa8ef9c2195158d8ff9bd0a737e56
but I also get 404 errors for these
https://testdata.mantidproject.org/ftp/external-data/MD5/0d2708e06bc1f086bd2f719c7142655c
https://testdata.mantidproject.org/ftp/external-data/MD5/69fa22403dbaff68d1f6d9b92c455030
Is it possible you didn't upload these two to the remote store?

Otherwise looks great to me, thanks so much for all your hard work!

andy-bridger commented:
I added all the files in one go, so maybe that was the problem, I've reuploaded them individually now

andy-bridger committed:
change fpath to forward slash?

RichardWaiteSTFC reviewed (approved):
Great work, thanks!
